src/Periphery/ReceiverStargateV2.sol

```diff
@@ -0,0 +1,239 @@
+}
+
+interface ILayerZeroComposer {
+    /**
+     * @notice Composes a LayerZero message from an OApp.
+     * @param _from The address initiating the composition, typically the OApp where the lzReceive was called.
+     * @param _guid The unique identifier for the corresponding LayerZero src/dst tx.
+     * @param _message The composed message payload in bytes. NOT necessarily the same payload passed via lzReceive.
+     * @param _executor The address of the executor for the composed message.
+     * @param _extraData Additional arbitrary data in bytes passed by the entity who executes the lzCompose.
+     */
+    function lzCompose(
+        address _from,
+        bytes32 _guid,
+        bytes calldata _message,
+        address _executor,
+        bytes calldata _extraData
+    ) external payable;
+}
+
+/// @title ReceiverStargateV2
+/// @author LI.FI (https://li.fi)
+/// @notice Arbitrary execution contract used for cross-chain swaps and message passing via Stargate V2
+/// @custom:version 1.0.0
+contract ReceiverStargateV2 is
+    ILiFi,
+    ReentrancyGuard,
+    TransferrableOwnership,
+    ILayerZeroComposer
+{
+    using SafeERC20 for IERC20;
+
+    /// Storage ///
+    IExecutor public immutable executor;
+    address public immutable endpointV2;
+    uint256 public immutable recoverGas;
+
+    /// Modifiers ///
+    modifier onlyEndpointV2() {
+        if (msg.sender != endpointV2) {
+            revert UnAuthorized();
+        }
+        _;
+    }
+
+    /// Constructor
+    constructor(
+        address _owner,
+        address _executor,
+        address _endpointV2,
+        uint256 _recoverGas
+    ) TransferrableOwnership(_owner) {
+        owner = _owner;
+        executor = IExecutor(_executor);
+        endpointV2 = _endpointV2;
+        recoverGas = _recoverGas;
+    }
+
+    /// External Methods ///
+
+    /// @notice Completes a stargateV2 cross-chain transaction on the receiving chain
+    /// @dev This function is called by Stargate Router via LayerZero endpoint (sendCompose(...) function)
+    /// @param _from The address initiating the composition, typically the OApp where the lzReceive was called
+    /// @param * (unused) The unique identifier for the corresponding LayerZero src/dst tx
+    /// @param _message The composed message payload in bytes. NOT necessarily the same payload passed via lzReceive
+    /// @param * (unused) The address of the executor for the composed message
+    /// @param * (unused) Additional arbitrary data in bytes passed by the entity who executes the lzCompose
+    function lzCompose(
+        address _from,
+        bytes32, // _guid (not used)
+        bytes calldata _message,
+        address, // _executor (not used)
+        bytes calldata // _extraData (not used)
+    )
+        external
+        payable
+        //nonReentrant // @REVIEWER: pls comment/confirm that we do not need this here (due to onlyEndpointV2 access control modifier)
```

Comment:
Reentrancy is checked in EndpointV2 itself.  In fact, this is Testable, although it is a bit difficult to do so.  Here is a rough example of how this is tested:

https://github.com/lifinance/contracts/compare/lf-8443-starcraftV2...ryandgoulding:contracts:lf-8443-starcraftV2__feedback?expand=1

Essentially, the above test delivers native eth to a contract with a `receive()` that has been contrived to make the re-entrant `lzCompose()` call.  It is then checked that the call fails with the reason LZ_COMPOSE_NOT_FOUND_SELECTOR and that the hash has already been seen before.

Reply:
awesome, I've added this test to our repo for reference and removed the nonReentrant modifier from the Receiver contract. Thank you for the feedback and the test case